### PR TITLE
SW-4610 Sort nurseries in withdrawal form, plumb SearchResponseElement type down through BatchWithdrawalFlow and SelectPurposeForm

### PR DESCRIPTION
**Pull request by nickgraz**

### CODE REVIEW

**Comment by nickgraz** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #1939** <a href="https://app.graphite.dev/github/pr/terraware/terraware-web/1939?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/terraware/terraware-web/1939?utm_source=stack-comment).

File: src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx [hunk 1]
```diff
@@ -377,20 +380,22 @@ export default function SelectPurposeForm(props: SelectPurposeFormProps): JSX.El
...
-          acc[batch.facility_id.toString()] = { label: batch.facility_name, value: batch.facility_id };
+      .reduce((acc: Record<string, DropdownItem>, batch: SearchResponseElement) => {
+        if (!acc[`${batch.facility_id}`]) {
+          acc[`${batch.facility_id}`] = { label: `${batch.facility_name}`, value: batch.facility_id };
         }
         return acc;
       }, {});
 
-    const options: DropdownItem[] = Object.values(nurseries);
+    const options: DropdownItem[] = Object.values(nurseries).sort((a, b) =>
```
Comment (nickgraz) [inline src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:396]: This is the only non-type-related change

**Comment by nickgraz** [comment]
I had to change a lot of things because I was moving away from the `any[]` type of the `batches` passed in. This might be better to hydrate the intended type in the parent component so we can pass a better type than `SearchResponseElement` down, but I didn't want to go more overboard than I already did. Let me know your thoughts

**Comment by karthikbtf** [comment]
> I had to change a lot of things because I was moving away from the `any[]` type of the `batches` passed in. This might be better to hydrate the intended type in the parent component so we can pass a better type than `SearchResponseElement` down, but I didn't want to go more overboard than I already did. Let me know your thoughts

Thanks for the changes. Yes, we need to move away from SearchResponseElement, for now replacing `any` with this type is good.

File: src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx [hunk 2]
```diff
@@ -248,7 +249,7 @@ export default function SelectPurposeForm(props: SelectPurposeFormProps): JSX.El
           setIndividualError('notReadyQuantityWithdrawn', strings.INVALID_VALUE);
           allValid = false;
         } else {
-          if (+notReadyQuantityWithdrawn > +batches[0]['notReadyQuantity(raw)']) {
+          if (+notReadyQuantityWithdrawn > Number(batches[0]['notReadyQuantity(raw)'])) {
```
Comment (karthikbtf) [inline src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:252]: might as well use Number on `notReadyQuantityWithdrawn` as well, a few other use cases below.

File: src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx [hunk 3]
```diff
@@ -351,22 +352,24 @@ export default function SelectPurposeForm(props: SelectPurposeFormProps): JSX.El
...
-          notReadyQuantityWithdrawn: isSingleOutplant ? 0 : isSingleBatch ? notReadyQuantityWithdrawn || 0 : 0,
-          readyQuantityWithdrawn: isSingleOutplant
-            ? withdrawnQuantity || 0
-            : isSingleBatch
-            ? readyQuantityWithdrawn || 0
-            : 0,
-        })),
+        .map(
+          (batch: SearchResponseElement): BatchWithdrawalPayload => ({
+            batchId: Number(batch.id),
```
Comment (karthikbtf) [inline src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:363]: Nice change, need to also update BatchWithdrawalFlow.tsxL129 (the comparison is failing, might need to convert them to Numbers before checking). The feature in question is this modal (which won't show in this PR but shows in staging). To trigger this modal, withdrawal all quantities as part of a withdrawal. 

Terraware App 2023-12-20 11-01-14

**Review by karthikbtf — COMMENTED** [review]
Looks good. A few comments.

**Review by karthikbtf — APPROVED** [review]
Thanks, nice changes.

File: src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx (hunk 3 above)
Comment (nickgraz) [reply src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:363]: Thanks, I'll double check this functionality once it hits staging

Comment (nickgraz) [reply src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:363]: Works as expected in staging

Comment (karthikbtf) [reply src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:363]: Awesome, thanks @nickgraz  🎉

**Pull request merged**